app/src/main/java/org/merakilearn/ui/onboarding/WelcomeFragment.kt

```diff
@@ -74,6 +80,63 @@ class WelcomeFragment : BaseFragment() {
         })
     }
 
+    private fun initGoogleSignInOption() {
+        val gso =
+            GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
+                .requestEmail()
+                .requestIdToken(getString(R.string.server_client_id))
+                .build()
+        mGoogleSignInClient = GoogleSignIn.getClient(requireActivity(), gso)
+    }
+
+    private fun signInWithGoogle() {
+        val signInIntent = mGoogleSignInClient!!.signInIntent
+        startActivityForResult(signInIntent, RC_SIGN_IN)
+    }
+
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+        if (requestCode == RC_SIGN_IN) {
+            val task =
+                GoogleSignIn.getSignedInAccountFromIntent(data)
+            handleSignInResult(task)
+        }
+    }
+
+    private fun handleSignInResult(completedTask: Task<GoogleSignInAccount>) {
+        try {
+            val account = completedTask.getResult(ApiException::class.java)
+            logOutSelectedAccount(account)
+        } catch (e: ApiException) {
+            Log.e(LoginFragment.TAG, "signInResult:failed code=", e)
+            toast(getString(R.string.unable_to_sign))
+        }
+    }
+
+    private fun logOutSelectedAccount(account: GoogleSignInAccount) {
+        mGoogleSignInClient?.signOut()?.addOnSuccessListener {
+            viewModel.idToken.value = account.idToken
+            observeFetchingBackendToken()
+        }
+    }
+
+    private fun observeFetchingBackendToken() {
+        toggleProgressBarVisibility(View.VISIBLE)
+        viewModel.loginResult.observe(this, Observer {
+            toggleProgressBarVisibility(View.GONE)
+            if (it) {
+                MainActivity.launch(requireContext())
```

Comment:
Better use events and state here. Also instead on calling launch, we can use navigator which already has a method for intent of home screen. Maybe add a method for launching the same in naviagator or just use the intent

Reply:
Launch is actually using intent only which is defined in activity companion object

Reply:
Yes still Welcome Fragment is using Main Activity explicitly. Its better if we use navigator for navigation in all new code we write atleast as it makes it easy for us to track all navigations or to make changes from a single source.

Reply:
I tried to implement this referring to method implemented already. I got confused in between. It would be great, if you can help with this one. It might give me insights bon MVI